src/features/shared/sliding/BaseSlidingVote.tsx

```diff
@@ -82,36 +124,162 @@ export default function BaseSlidingVote({
     [dispatch, isPost, item, present, presentLoginIfNeeded]
   );
 
-  const startActions: [SlidingItemAction, SlidingItemAction] = useMemo(() => {
-    return [
-      {
-        render: () => (
-          <VoteArrow
-            slash={currentVote === 1}
-            bgColor="primary"
-            icon={arrowUpSharp}
-          />
-        ),
+  const reply = useCallback(async () => {
+    if (presentLoginIfNeeded()) return;
+    const reply = await presentCommentReply(item);
+    if (!isPost && reply) prependComments([reply]);
+  }, [
+    item,
+    isPost,
+    presentCommentReply,
+    presentLoginIfNeeded,
+    prependComments,
+  ]);
+
+  const { id, isSaved } = useMemo(() => {
+    if (isPost) {
+      const id = item.post.id;
+      return { id: id, isSaved: postSavedById[id] };
+    } else {
+      const id = item.comment.id;
+      return { id: id, isSaved: commentSavedById[id] };
+    }
+  }, [item, isPost, postSavedById, commentSavedById]);
+
+  const save = useCallback(async () => {
+    if (presentLoginIfNeeded()) return;
+    try {
+      await dispatch((isPost ? savePost : saveComment)(id, !isSaved));
+    } catch (error) {
+      present({
+        message: "Failed to mark item as saved",
+        duration: 3500,
+        position: "bottom",
+        color: "danger",
+      });
+      throw error;
+    }
+  }, [presentLoginIfNeeded, dispatch, isPost, id, isSaved, present]);
+
+  const saveAction = useMemo(() => {
+    return {
+      icon: bookmarkOutline,
+      trigger: save,
+      bgColor: "success",
+      slash: isSaved,
+    };
+  }, [save, isSaved]);
+
+  const hideAction = useMemo(() => {
+    return onHide
+      ? {
+          icon: isHidden ? eyeOutline : eyeOffOutline,
+          trigger: () => {
+            if (presentLoginIfNeeded()) return;
+            onHide();
+          },
+          bgColor: isHidden ? "tertiary" : "danger",
+        }
+      : undefined;
+  }, [presentLoginIfNeeded, isHidden, onHide]);
+
+  const collapseRootComment = useCollapseRootComment(
+    !isPost ? item : undefined,
+    rootIndex
+  );
+  const collapseAction = useMemo(() => {
+    return collapseRootComment
+      ? {
+          icon: collapsed ? chevronExpand : chevronCollapse,
+          trigger: collapseRootComment,
+          bgColor: "tertiary",
+        }
+      : undefined;
+  }, [collapsed, collapseRootComment]);
+
+  const isRead = useMemo(() => {
+    return readByInboxItemId[getInboxItemId(item as InboxItemView)] ?? false;
+  }, [item, readByInboxItemId]);
+
+  const markUnread = useCallback(async () => {
+    try {
+      await dispatch(markRead(item as InboxItemView, !isRead));
```

Comment:
Refactor to not use `as`

Reply:
How should I do that?

Reply:
Something like this would work, then you can remove the changes to inboxSlice:

```diff
--- a/src/features/shared/sliding/BaseSlidingVote.tsx
+++ b/src/features/shared/sliding/BaseSlidingVote.tsx
@@ -30,7 +30,6 @@ import { PageContext } from "../../auth/PageContext";
 import { SwipeAction, SwipeActions } from "../../../services/db";
 import useCollapseRootComment from "../../comment/useCollapseRootComment";
 import { getInboxItemId, markRead } from "../../inbox/inboxSlice";
-import { InboxItemView } from "../../inbox/InboxItem";
 import { CommentsContext } from "../../comment/CommentsContext";
 import styled from "@emotion/styled";

@@ -38,10 +37,25 @@ const StyledItemContainer = styled.div`
   --ion-item-border-color: transparent;
 `;

+type SlideableItem =
+  | CommentView
+  | PostView
+  | PersonMentionView
+  | CommentReplyView;
+
+function isInboxItem(
+  item: SlideableItem
+): item is PersonMentionView | CommentReplyView {
+  if ("person_mention" in item) return true;
+  if ("comment_reply" in item) return true;
+
+  return false;
+}
+
 interface BaseSlidingVoteProps {
   children: React.ReactNode;
   className?: string;
-  item: CommentView | PostView | PersonMentionView | CommentReplyView;
+  item: SlideableItem;
   rootIndex?: number;
   collapsed?: boolean;
   actions: SwipeActions;
@@ -198,12 +212,14 @@ function BaseSlidingVoteInternal({
   }, [collapsed, collapseRootComment]);

   const isRead = useMemo(() => {
-    return readByInboxItemId[getInboxItemId(item as InboxItemView)] ?? false;
+    return isInboxItem(item) ? readByInboxItemId[getInboxItemId(item)] : false;
   }, [item, readByInboxItemId]);

   const markUnread = useCallback(async () => {
+    if (!isInboxItem(item)) return;
+
     try {
-      await dispatch(markRead(item as InboxItemView, !isRead));
+      await dispatch(markRead(item, !isRead));
     } catch (error) {
       present({
         message: "Failed to mark item as unread",
```